DOCS-915: Document entry point file for modules and improve modules documentation (#1476)
viamrobotics/docs: docs/extend/modular-resources/create.md

title: "Code your own modules to create custom resources"
linkTitle: "Create"
weight: 20
type: "docs"
tags: ["server", "rdk", "extending viam", "modular resources", "components", "services"]
description: "Use the Viam module system to implement custom resources that can be included in any Viam-powered robot."
no_list: true
---

The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}}, [components](/components/), and [services](/services/), into any robot running on Viam.

A common use case for modular resources is to create a new model that implements an existing Viam API.
However, you can also create and expose new API types using modular resources.

{{% alert title="Modules vs. modular resources" color="tip" %}}

A configured *module* can make one or more *modular resources* available for configuration.

{{% /alert %}}

## Create a custom modular resource

To create your own modular resource, code a module in Go or Python using the module support libraries provided by [Viam's SDKs](/program/apis/) that implements at least one new model or type of {{< glossary_tooltip term_id="resource" text="resource" >}}:

{{< tabs >}}

> {{% tab name="Define a New Model of a Built-In Resource Type" %}}
> 

Reviewer: ```suggestion
{{% tab name="New Model" %}}
Define a new model of a built-in resource type:

```
[opt] Realized these tab names are so long that they show up stacked oddly on many screens.

1. [Code a new resource model](#code-a-new-resource-model) implementing all methods the Viam RDK requires in the API definition of its built-in type (ex. `rdk:component:base`).
Import your custom model and API into the main program and register the new resource model with your chosen SDK's helper methods.

2. [Code a main program](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) that starts the module after adding your desired resources from the registry.
This main program is the "entry point" to your module.

3. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.

{{% /tab %}}

> {{% tab name="Define a New Type of Resource" %}}
> 

Reviewer: ```suggestion
{{% tab name="New Type" %}}
Define a new type of resource:

```
see above, opt

2. [Code at least one model](#code-a-new-resource-model) of this new resource.
Make sure to implement every method required in your API definition.
Import your custom models and APIs into the main program and register the new resource models with your chosen SDK's helper methods.

3. [Code a main program](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) that starts the module after adding your desired resources from the registry.
This main program is the "entry point" to your module.

4. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.

{{% /tab %}}
{{% /tabs %}}

### Code a new resource model

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/extend/modular-resources/key-concepts/#models) as a new model, `"mybase"`:

> The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).

Reviewer: ```suggestion
The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) are adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```

Reviewer: Oh man my suggestion is grammatically incorrect sorry!

[Go Client SDK](https://pkg.go.dev/go.viam.com/rdk):

<file>mybase.go</file>

> {{%expand "Click to view example code from a Go module implementing a custom model of the base component built-in resource, mybase, and registering the new model and API" %}}

Reviewer: ```suggestion
{{%expand "Click to view example code from a Go module implementing a custom model of the base component built-in resource, mybase, and registering the new model" %}}
```
I think this could confuse people and make people think they're creating a new API instead of implementing an existing one

Author: Maybe should add the helpers clarification?

Author: like 
```suggestion
{{%expand "Click to view example code from a Go module implementing a custom model of the base component built-in resource, mybase, and registering the new model and API helpers with the client SDK" %}}
```
Struggling as it's a little long but lmk what you think. can incorporate this feedback either way when I get time to rework the whole section re feedback too

Reviewer: I'm not sure what the "helpers" adds? I think shorten it as much as possible, and if you have to you can include other details inside the expander or in the body text. See also my formatting suggestion though. And, if you come up with an entirely different format, this could become moot anyway? But if not, maybe:
```suggestion
{{%expand "Click to view example code from a Go module implementing and registering a custom model of the base component" %}}
```

Author: done, and added details!

        if abs(linear.y) < 0.01 and abs(angular.z) < 0.01:
            return self.stop(extra=extra, timeout=timeout)

        # use linear and angular velocity to calculate percentage of max power to pass to SetPower for left & right motors
        sum = abs(linear.y) + abs(angular.z)

        self.left.set_power(power=((linear.y - angular.z) / sum), extra=extra, timeout=timeout)
        self.right.set_power(power=((linear.y + angular.z) / sum), extra=extra, timeout=timeout)

    # stop: stop the base from moving by stopping both motors
    async def stop(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs):
        self.left.stop(extra=extra, timeout=timeout)
        self.right.stop(extra=extra, timeout=timeout)

    # is_moving: check if either motor on the base is moving with motors' is_powered
    async def is_moving(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> bool:
        return self.left.is_powered(extra=extra, timeout=timeout)[0] or self.right.is_powered(extra=extra, timeout=timeout)[0]
```

{{% /expand%}}

<file>_\_init__.py</file>

> {{%expand "Click to view example code from a Python module registering the mybase custom model and API" %}}

Reviewer: ```suggestion
{{%expand "Click to view example code from a Python module registering the mybase custom model" %}}
```
See above

Author: changed to details and took that out of the clicker

``` python {class="line-numbers linkable-line-numbers"}
"""
This file registers the MyBase model with the Python SDK.
"""

from viam.components.motor import *
from viam.components.base import Base
from viam.resource.registry import Registry, ResourceCreatorRegistration

from .my_base import MyBase

Registry.register_resource_creator(Base.SUBTYPE, MyBase.MODEL, ResourceCreatorRegistration(MyBase.new_base, MyBase.validate_config))
```

{{% /expand%}}

### Code a main program that serves as the entry point file to the module

[Go Client SDK](https://pkg.go.dev/go.viam.com/rdk):

<file>main.go</file>

> {{%expand "Click to view example code from a Go module's entry point file initializing the mybase custom model and API from the registry" %}}

Reviewer: ```suggestion
{{%expand "Click to view example code from a Go module's entry point file initializing the mybase custom model from the registry" %}}
```
As above--are they actually registering an API or just a model that uses that API?

Author: registering helpers w SDK for both I think, it's confusing

Reviewer: Hey for the sake of clarity etc. instead of editing my suggestions themselves, mind making a new suggestion on the thread and copy pasting from mine if you want?

What is an API helper? I'm still thinking "API" should come out of here

Reviewer: But I do like taking "from the registry" off too since it's long

import (
    "context"

    "github.com/edaniels/golog"

    "go.viam.com/rdk/components/base"
    "go.viam.com/rdk/module"
    "go.viam.com/utils"

    // NOTE: You must update the following line to import your local package "mybase"
    "go.viam.com/rdk/examples/customresources/models/mybase"
)

func main() {
    // NewLoggerFromArgs will create a golog.Logger at "DebugLevel" if
    // "--log-level=debug" is an argument in os.Args and at "InfoLevel" otherwise.
    utils.ContextualMain(mainWithArgs, module.NewLoggerFromArgs("yourmodule"))
}

func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) (err error) {
    myMod, err := module.NewModuleFromArgs(ctx, logger)
    if err != nil {
        return err
    }

    // Models and APIs add helpers to the registry during their init().
    // They can then be added to the module here.
    err = myMod.AddModelFromRegistry(ctx, base.API, mybase.Model)
    if err != nil {
        return err
    }

    err = myMod.Start(ctx)
    defer myMod.Close(ctx)
    if err != nil {
        return err
    }
    <-ctx.Done()
    return nil
}
```

{{% /expand%}}

[Python SDK](https://python.viam.dev/):

<file>main.py</file>

> {{%expand "Click to view example code from a Python module's entry point file initializing the mybase custom model and API from the registry" %}}

Reviewer: ```suggestion
{{%expand "Click to view example code from a Python module's entry point file initializing the mybase custom model from the registry" %}}
```
see above

- Run `ls` in your module directory to find the executable, which should have been named after the module directory.

Expand the [Go module code](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) to view <file>main.go</file> for an example of this.
<file>main.go</file> adds the custom model <file>mybase.go</file> from the resource registry, while <file>mybase.go</file> defines and registers the module.

{{% /tab %}}
{{% tab name="Python" %}}

If using the Python SDK, one option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.

For example:

``` shell
#!/bin/sh
cd `dirname $0`

# Be sure to use `exec` so that termination signals reach the python process,
# or handle forwarding termination signals manually
exec python3 -m <your-module-directory-name>.<main-program-filename-without-extension> $@
```

To make this shell script executable, run the following command in your terminal:

> ``` shell

Reviewer: ```suggestion
```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
```
To make it have a black background and more clearly be a terminal command

Author: done

sudo chmod +x <FILEPATH>/<FILENAME>
```

{{% /tab %}}

> {{% /tabs %}}

Reviewer: ```suggestion
{{< /tabs >}}
```
[nit] Just because the opening one is like this, so makes sense to match them I think

You need to ensure any dependencies for your module (including the Viam SDK) are installed, as well.
Your executable will be run by `viam-server` as root, so dependencies need to be available to the root user.

## Next steps

> Follow [these instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.
> 
> Make your [executable accessible](/extend/modular-resources/configure/#make-sure-viam-server-can-access-your-executable), [configure your new module](/extend/modular-resources/configure/#configure-your-module), and [configure your modular resource](/extend/modular-resources/configure/#configure-your-modular-resource).

Reviewer: ```suggestion
Follow [these configuration instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.
```
I think it's best to leave it at that so people don't feel the need to click all the links separately or fear they're missing something, when one link has everything they need to do.

{{% /tab %}}
{{% tab name="Define a New Type of Resource" %}}

1. Define the messages and methods of the new API in [protobuf](https://github.com/protocolbuffers/protobuf), then [generate code](https://grpc.io/docs/languages/python/generated-code/) in Python or Go and use the generated code to implement the higher level server and client functions required.

2. [Code at least one model](#code-a-new-resource-model) of this new resource.
Make sure to implement every method required in your API definition.
Import your custom models and APIs into the main program and register the new resource models with your chosen SDK's helper methods.

3. [Code a main program](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) that starts the module after adding your desired resources from the registry.
This main program is the "entry point" to your module.

4. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.

{{% /tab %}}
{{% /tabs %}}

### Code a new resource model

> 

Author: ```suggestion
```

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/extend/modular-resources/key-concepts/#models) as a new model, `"mybase"`:

The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).

> 

Author: ```suggestion
```

---
title: "Code your own modules to create custom resources"
linkTitle: "Create"
weight: 20
type: "docs"
tags: ["server", "rdk", "extending viam", "modular resources", "components", "services"]
description: "Use the Viam module system to implement custom resources that can be included in any Viam-powered robot."
no_list: true
---

The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}}, [components](/components/), and [services](/services/), into any robot running on Viam.

> A common use-case for modular resources is to create a new model using an existing Viam API.

Reviewer: ```suggestion
A common use case for modular resources is to create a new model that implements an existing Viam API.
```
"use case" is correct. The other part is [opt].

{{% /tab %}}
{{% tab name="Define a New Type of Resource" %}}

1. Define the messages and methods of the new API in [protobuf](https://github.com/protocolbuffers/protobuf), then [generate code](https://grpc.io/docs/languages/python/generated-code/) in Python or Go and use the generated code to implement the higher level server and client functions required.

2. [Code at least one model](#code-a-new-resource-model) of this new resource.
Make sure to implement every method required in your API definition.
Import your custom models and APIs into the main program and register the new resource models with your chosen SDK's helper methods.

3. [Code a main program](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) that starts the module after adding your desired resources from the registry.
This main program is the "entry point" to your module.

4. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.

{{% /tab %}}
{{% /tabs %}}

> ### Code a new resource model

Reviewer: I like how you've created headings that match the steps described in the overview just above. However, what do people do for step one of define a new type of resource? Is there a way to point them to how to do that, if it's not covered right here?

Reviewer: Oh, 762? Okay fair enough!

> {{% alert title="Note" color="note" %}}

Reviewer: I don't think this needs to be a note; it is just body text I think. Also, we're moving away from "note" to "important" or "stability notice" (and you can still use "tip" or "info" in green as well.

> {{% alert title="Note" color="note" %}}

Author: ```suggestion
```

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/extend/modular-resources/key-concepts/#models) as a new model, `"mybase"`:

The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).

> {{% /alert %}}

Author: ```suggestion
```

> [Go Client SDK](https://pkg.go.dev/go.viam.com/rdk):

Reviewer: I think the format of this section is a bit difficult to parse visually because our expanders have bold text, whereas the surrounding "headings" aren't headings so the Python vs Go and the file names get lost visually. Maybe try using tabs like with the existing doc, and/or maybe put in horizontal lines to break up the sections? Or use headings or something? Something to emphasize the desired hierarchy/sections instead of the expanders being the biggest boldest thing.

Author: I agree. I tried to put the code in tabs but the expanders don't work inside of tabs and the code is chunky. Unsure @JessamyT or @npentrel do you have any ideas on this? maybe the lines?

Reviewer: TBH I found the old format with the tabs and no expanders easier to follow/look at. Yeah I'd play with the "---" line separators perhaps. Maybe I'll mess around locally and make a suggestion if I come up with anything I actually like.

Reviewer: Ooh another thing: since we do Python as the front tab, then Go, on most other pages, maybe do that here?

Reviewer: Okay so, if you want to stick with this general organizational structure (as opposed to the way it was before) but use tabs, you could use tabs with the "details" tag like https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab and those do indeed nest within tabs.

Reviewer: Also, I'm thinking you don't really need the link to the Go or Python SDKs on these "headings"

Reviewer: With details tags, it looks like this: 
![image](https://github.com/viamrobotics/docs/assets/75634662/56fdae9d-dc13-4d1d-b723-3c450939ec36)
Here's a gist incorporating all my feedback on this section (does not include executable and next steps):
https://gist.github.com/JessamyT/514cd1da1694ade1bdff0a96a57bcda0

Expand the [Go module code](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) to view <file>main.go</file> for an example of this.
<file>main.go</file> adds the custom model <file>mybase.go</file> from the resource registry, while <file>mybase.go</file> defines and registers the module.

{{% /tab %}}
{{% tab name="Python" %}}

If using the Python SDK, one option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.

For example:

``` shell
#!/bin/sh
cd `dirname $0`

# Be sure to use `exec` so that termination signals reach the python process,
# or handle forwarding termination signals manually
exec python3 -m <your-module-directory-name>.<main-program-filename-without-extension> $@
```

To make this shell script executable, run the following command in your terminal:

``` shell
sudo chmod +x <FILEPATH>/<FILENAME>
```

{{% /tab %}}
{{% /tabs %}}

You need to ensure any dependencies for your module (including the Viam SDK) are installed, as well.
Your executable will be run by `viam-server` as root, so dependencies need to be available to the root user.

> ## Next Steps

Reviewer: ```suggestion
## Next steps
```
[nit] The other headings are sentence case

Follow [these instructions](/extend/modular-resources/configure/) to add your custom resource to your robot.

> Make your [executable accessible](/extend/modular-resources/configure/#make-sure-viam-server-can-access-your-executable), [configure your new module](/extend/modular-resources/configure/#configure-your-module), and [configure your modular resource](/extend/modular-resources/configure/#configure-your-modular-resource).

Reviewer: ```suggestion
```
This page just had them do all this so I don't think it makes sense to put this in next steps.

Author: I think it helps to provide an overview of what they haven't done yet? After completing the create page they haven't done that

Reviewer: Ahh sorry I misread. You're totally right. However, I still think it's unnecessary to list this all out with links since if they click the link above it takes them to all of this. Will make a different suggestion.

> 
> After configuring a modular resource, test it with the [Control tab](/manage/fleet/#remote-control) and program it with the [Viam SDKs](/program/apis/).

Reviewer: ```suggestion
```
I think one next step (configure) makes sense here, and for clarity it's best to just leave it at that. Maybe add this step to the bottom of the configure page though!! I think that'd be great!

Author: Same as above

Reviewer: I'm saying you should move this line to the bottom of the configure instructions since it's not relevant to people here yet, until after they configure. I think of the "Next steps" as being an actionable place where we guide them to exactly what to do next, and I don't want them to go try and test stuff when they can't yet.

    err = myMod.Start(ctx)
    defer myMod.Close(ctx)
    if err != nil {
        return err
    }
    <-ctx.Done()
    return nil
}
```

</details>

{{% /tab %}}
{{< /tabs >}}

{{% alert title="Important" color="note" %}}

You must define all functions belonging to a built-in resource subtype's API if defining a new model.
Otherwise, the class won’t instantiate.

- If you are using the Python SDK, raise an `NotImplementedError()` in the body of functions you do not want to implement or put `pass`.
- If you are using the Go SDK, return `errUnimplemented`.
- Additionally, return any values designated in the function's return signature, typed correctly.

{{% /alert %}}

### Compile the module into an executable

To [add a module](/extend/modular-resources/configure/#configure-your-module) to the configuration of your robot, you need to have an [executable](https://en.wikipedia.org/wiki/Executable) that runs your module when executed, can take a local socket as a command line argument, and cleanly exits when sent a termination signal.

Your options for completing this step are flexible, as this file does not need to be in a raw binary format.

{{% tabs %}}
{{% tab name="Python" %}}

> If using the Python SDK, one option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.

Author: ```suggestion
One option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.
```

Author: @JessamyT with this suggestion it might address your concern but keep the wording ambiguous?

Reviewer: @sguequierre The goal is to not be ambiguous :D But okay I guess there's the mention of flexibility just above the tabs. To make it less ambiguous you could say "The simplest option is to create and save..." instead of "One option..." but not going to block on this

---
title: "Code your own modules to create custom resources"
linkTitle: "Create"
weight: 20
type: "docs"
tags: ["server", "rdk", "extending viam", "modular resources", "components", "services"]
description: "Use the Viam module system to implement custom resources that can be included in any Viam-powered robot."
no_list: true
---

> The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}}, [components](/components/), and [services](/services/), into any robot running on Viam.

Reviewer: ```suggestion
The Viam module system allows you to integrate custom [components](/components/) and [services](/services/) into any robot running on Viam.
```
Or leave resources and take out components and services, or do 
```suggestion
The Viam module system allows you to integrate custom {{< glossary_tooltip term_id="resource" text="resources" >}} ([components](/components/) and [services](/services/)) into any robot running on Viam.
```

Import your custom models and APIs into the main program and register the new resource models with your chosen SDK's helper methods.

1. [Code a main program](#code-a-main-program-that-serves-as-the-entry-point-file-to-the-module) that starts the module after adding your desired resources from the registry.
This main program is the "entry point" to your module.

1. [Compile or package](#compile-the-module-into-an-executable) the module into a single executable that can receive a socket argument from Viam, open the socket, and start the module at the entry point.

{{% /tab %}}
{{% /tabs %}}

### Code a new resource model

The following example module registers a modular resource implementing Viam's built-in [Base API](/components/base/#api) [(rdk:service:base)](/extend/modular-resources/key-concepts/#models) as a new model, `"mybase"`:

The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) are adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).

{{< tabs name="Sample SDK Code">}}

> {{% tab name="Go"%}}

Reviewer: Is there a reason for showing Go before Python on all of these, contrary to what we do in the rest of the docs?

Author: No just switched up in my head, fixed!

        # stop the base if absolute value of linear and angular velocity is less than .01
        if abs(linear.y) < 0.01 and abs(angular.z) < 0.01:
            return self.stop(extra=extra, timeout=timeout)

        # use linear and angular velocity to calculate percentage of max power to pass to SetPower for left & right motors
        sum = abs(linear.y) + abs(angular.z)

        self.left.set_power(power=((linear.y - angular.z) / sum), extra=extra, timeout=timeout)
        self.right.set_power(power=((linear.y + angular.z) / sum), extra=extra, timeout=timeout)

    # stop: stop the base from moving by stopping both motors
    async def stop(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs):
        self.left.stop(extra=extra, timeout=timeout)
        self.right.stop(extra=extra, timeout=timeout)

    # is_moving: check if either motor on the base is moving with motors' is_powered
    async def is_moving(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> bool:
        return self.left.is_powered(extra=extra, timeout=timeout)[0] or self.right.is_powered(extra=extra, timeout=timeout)[0]
```

> <br>
> </details>
> 

Reviewer: ```suggestion

</details>
<br>

```

> <file>__init__.py</file> registers the mybase custom model and API helper functions with the SDK.

Reviewer: right now in the rendering, init is showing up bold instead of with the underscores.

    # is_moving: check if either motor on the base is moving with motors' is_powered
    async def is_moving(self, *, extra: Optional[Dict[str, Any]] = None, timeout: Optional[float] = None, **kwargs) -> bool:
        return self.left.is_powered(extra=extra, timeout=timeout)[0] or self.right.is_powered(extra=extra, timeout=timeout)[0]
```

<br>
</details>

<file>__init__.py</file> registers the mybase custom model and API helper functions with the SDK.

<details>
  <summary>Click to view sample code from <file>__init__.py</file></summary>

``` python {class="line-numbers linkable-line-numbers"}
"""
This file registers the MyBase model with the Python SDK.
"""

from viam.components.motor import *
from viam.components.base import Base
from viam.resource.registry import Registry, ResourceCreatorRegistration

from .my_base import MyBase

Registry.register_resource_creator(Base.SUBTYPE, MyBase.MODEL, ResourceCreatorRegistration(MyBase.new_base, MyBase.validate_config))
```

</details>

{{% /tab %}}
{{< /tabs >}}

### Code a main program that serves as the entry point file to the module

{{< tabs name="Sample SDK Main Program Code">}}
{{% tab name="Go"%}}

> <file>main.go</file> is the demo Go module's entry point file.

Reviewer: ```suggestion
<file>main.go</file> is the Go module's entry point file.
```
I think reads better

Author: ok!

> When executed, it initializes the mybase custom model and API helper functions from the registry.

Reviewer: ```suggestion
When executed, it initializes the `mybase` custom model.
```
I think the way the custom model uses the API *is* the custom model, so I think this can be shortened.

    "go.viam.com/rdk/components/base"
    "go.viam.com/rdk/module"
    "go.viam.com/utils"

    // NOTE: You must update the following line to import your local package "mybase"
    "go.viam.com/rdk/examples/customresources/models/mybase"
)

func main() {
    // NewLoggerFromArgs will create a golog.Logger at "DebugLevel" if
    // "--log-level=debug" is an argument in os.Args and at "InfoLevel" otherwise.
    utils.ContextualMain(mainWithArgs, module.NewLoggerFromArgs("yourmodule"))
}

func mainWithArgs(ctx context.Context, args []string, logger golog.Logger) (err error) {
    myMod, err := module.NewModuleFromArgs(ctx, logger)
    if err != nil {
        return err
    }

    // Models and APIs add helpers to the registry during their init().
    // They can then be added to the module here.
    err = myMod.AddModelFromRegistry(ctx, base.API, mybase.Model)
    if err != nil {
        return err
    }

    err = myMod.Start(ctx)
    defer myMod.Close(ctx)
    if err != nil {
        return err
    }
    <-ctx.Done()
    return nil
}
```

</details>

{{% /tab %}}
{{% tab name="Python"%}}

<file>main.py</file> is the demo Python module's entry point file.

> When executed, it initializes the mybase custom model and API helper functions from the registry.

Reviewer: ```suggestion
When executed, it initializes the `mybase` custom model.
```
see above; I think this is redundant?

    err = myMod.Start(ctx)
    defer myMod.Close(ctx)
    if err != nil {
        return err
    }
    <-ctx.Done()
    return nil
}
```

</details>

{{% /tab %}}
{{< /tabs >}}

{{% alert title="Important" color="note" %}}

You must define all functions belonging to a built-in resource subtype's API if defining a new model.
Otherwise, the class won’t instantiate.

- If you are using the Python SDK, raise an `NotImplementedError()` in the body of functions you do not want to implement or put `pass`.
- If you are using the Go SDK, return `errUnimplemented`.
- Additionally, return any values designated in the function's return signature, typed correctly.

{{% /alert %}}

### Compile the module into an executable

To [add a module](/extend/modular-resources/configure/#configure-your-module) to the configuration of your robot, you need to have an [executable](https://en.wikipedia.org/wiki/Executable) that runs your module when executed, can take a local socket as a command line argument, and cleanly exits when sent a termination signal.

Your options for completing this step are flexible, as this file does not need to be in a raw binary format.

{{% tabs %}}
{{% tab name="Python" %}}

> If using the Python SDK, one option is to create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.

Reviewer: ```suggestion
Create and save a new shell script (<file>.sh</file>) that runs your module at your entry point (main program) file.
```
They're already on the Python tab so you can take that part out. What is the other option if not this option? Either say what it is or take that out I think.

Author: I was working on this section refactor with james quite a bit recently, the diff would just be some words if I hadn't moved the text from a different page.  In my understanding the wording is intentional bc there are a lot of different ways you can set the module executable. This is what James suggested as simplest use case for each sdk

For example:

``` sh {id="terminal-prompt" class="command-line" data-prompt="$"}
#!/bin/sh
cd `dirname $0`

# Be sure to use `exec` so that termination signals reach the python process,
# or handle forwarding termination signals manually
exec python3 -m <your-module-directory-name>.<main-program-filename-without-extension> $@
```

To make this shell script executable, run the following command in your terminal:

``` sh {id="terminal-prompt" class="command-line" data-prompt="$"}
sudo chmod +x <FILEPATH>/<FILENAME>
```

{{% /tab %}}
{{% tab name="Go" %}}

> If using the Go SDK, use Go to [compile](https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies) and obtain a single executable for your module:

Reviewer: ```suggestion
Use Go to [compile](https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies) and obtain a single executable for your module:
```

> The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) are adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).

Reviewer: ```suggestion
The Go code for the custom model (<file>mybase.go</file>) and module entry point file (<file>main.go</file>) is adapted from the full demo modules available on the [Viam GitHub](https://github.com/viamrobotics/rdk/blob/main/examples/customresources).
```
sorry I realize I accidentally suggested the opposite earlier

Author: fixed
